Extend wait_for on workflow context to take multiple futures and a condition block

jeffschoner-stripe opened the pull request:
### Summary

Extends the `wait_for` method on workflow context to take multiple futures and a condition block. This provides the equivalent to [`Workflow.await` in Java](https://github.com/temporalio/sdk-java/blob/master/temporal-sdk/src/main/java/io/temporal/workflow/Workflow.java#L822) and [`workflow.Await` in Go](https://pkg.go.dev/go.temporal.io/sdk@v1.11.0/workflow#Await) and a "wait for any" function. This function blocks workflow progress until one of the futures is finished or the block returns true. It can be used for a variety of cases where a workflow needs to wait for the completion of any combination of activities or timer completion, or receipt of a signal.

Also included:
- A new `WaitForWorkflow` example that demonstrates all three of these cases, along with an integration test to run it and verify its behavior
- Support in local workflow context testing mode
- Unit tests for changes to the dispatcher

### Testing

New new dispatcher unit tests: `rspec spec/unit/lib/temporal/workflow/dispatcher_spec.rb:78`
New integration test: `pushd examples && rspec spec/integration/wait_for_workflow_spec.rb && popd` (need to run `pushd examples && bin/worker` in another terminal)

antstorm commented on spec/unit/lib/temporal/testing/local_workflow_context_spec.rb:
```diff
@@ -126,4 +126,23 @@ def execute
     workflow_context.execute_activity!(TestHeartbeatingActivity)
   end
+
+  it 'can await with false condition' do
```
nit: this needs to be in `describe '#await' do` block for clarify and to match our current testing structure

antstorm reviewed (changes requested):
Thanks for the PR, @jeffschoner-stripe, this is a really cool one!

I left a few inline comments. Couple bigger points to think about:

1. I understand that `await` matches Java & Go SDKs, but it doesn't match our `wait_for` and `wait_for_all`. Not necessarily a problem, but wonder how you think about it
2. We can probably add other argument types to `await` (just a straight boolean and a future), it doesn't always have to be a block and can then replace `wait_for`. Doesn't have to be this PR, but let's make sure APIs are compatible
3. We need to make the example a bit clearer, maybe we can break it down into smaller methods? Plus a few inline comments on it
4. Dispatcher changes

Please let me know if all that makes sense

jeffschoner-stripe commented on examples/workflows/await_workflow.rb:
```diff
@@ -0,0 +1,64 @@
+
+      future = EchoActivity.execute("hi #{i}")
+      future.done do
+        activity_futures.delete(i)
```
I wasn't sure what to do about this. It is a race condition in threaded code, but it's not based on how workflow code is currently executed. Ideas on how to make this more proper? I don't have much expertise with multi-threaded Ruby code.

antstorm replied:
Just move the line `activity_futures[I] = future` to happen before you're setting the callback :) The reason it's not causing a race is because `EchoActivity.execute` doesn't start the activity straight away, but rather at the next blocking point

jeffschoner-stripe replied:
Got it now. I think as written, it would fail if run in unit test mode, although it does in practice work fine running against a Temporal server.

antstorm replied:
Yes, it works against the Temporal server because the commands are never dispatched before a blocking call (in this case it would be the next iteration's `.await`). But still a good idea to keep these things free from races

jeffschoner-stripe commented on lib/temporal/workflow/dispatcher.rb:
```diff
@@ -5,21 +5,41 @@ class Dispatcher
         handlers[target] << [event_name, handler]
       end
 
+      def register_await_handler(&handler)
```
The function is used by the workflow context to register a global, non-target handler. I could refactor this into `register_handler` where, say, if `target` is `nil` we treat it as a wildcard that affects all targets. What do you think about that approach?

antstorm replied:
Oh, I see, I think I mixed local and main contexts. The approach for using a wildcard handler target works 👍

jeffschoner-stripe commented on lib/temporal/workflow/dispatcher.rb:
```diff
@@ -5,21 +5,41 @@ class Dispatcher
+      def register_await_handler(&handler)
+        # This should only happen in situations where multithreading is being used
+        # in workflow code which is unsupported.
+        raise 'Await handler already active' unless await_handler.nil?
```
I'll add a test for this. I originally had an array of await handlers, but convinced myself there should only be one active at a time. It'll be easy to restore that original approach, but I need to make sure the behavior is as expected.

antstorm replied:
Yeah, I know what you mean. When programming a reactor you either need to disable certain behaviours or be prepared for any amount of context entanglement. I found the latter to be the simplest and most elegant solution

jeffschoner-stripe commented on lib/temporal/workflow/dispatcher.rb:
```diff
@@ -5,21 +5,41 @@ class Dispatcher
+        # Invoking the await handler needs to be done after target-specific handlers
+        # because an activity, sleep, or signal may have completed that affects the state
+        # evaluated in an await condition.
+        await_handler.call(unset_await_handler)
```
As far as I can tell, I need to check all await conditions whenever something potentially being waited on has completed (signal, activity, timer). I'm not sure how to do this without hooking into the dispatcher, as this is the central place where those completions trigger callbacks.

jeffschoner-stripe commented:
> Thanks for the PR, @jeffschoner-stripe, this is a really cool one!
> 
> I left a few inline comments. Couple bigger points to think about:
> 
> 1. I understand that `await` matches Java & Go SDKs, but it doesn't match our `wait_for` and `wait_for_all`. Not necessarily a problem, but wonder how you think about it

Would you prefer that I merge this logic into `wait_for` where it behaves this wait if a block is passed? Or maybe naming it something like `wait_until`?

> 2. We can probably add other argument types to `await` (just a straight boolean and a future), it doesn't always have to be a block and can then replace `wait_for`. Doesn't have to be this PR, but let's make sure APIs are compatible
> 3. We need to make the example a bit clearer, maybe we can break it down into smaller methods? Plus a few inline comments on it
> 4. Dispatcher changes
> 
> Please let me know if all that makes sense

These make sense. I'll work on some revisions.

jeffschoner-stripe commented:
@antstorm I still have more to do, but take a look at how I've refactored the dispatcher to be more generic, as I think this is the most fundamental change requested.

I've tried testing an await within a signal block, and the fiber seems to just end up never being resumed. I still need to experiment more with this. I've seen similar problems synchronously starting activities (i.e., waiting on a future) within a signal handler. Folks at Temporal have recommended against doing this generally, but I need to figure out if I can fix this, or at least have confidence I'm making this no worse than the current state.

antstorm commented:
@jeffschoner-stripe thank you for the changes, I really like how it's shaping up. Please let me know if you need help with figuring out the issues you've mentioned.

jeffschoner-stripe commented:
@antstorm, I believe I've addressed all your feedback now

I merged `await` into `wait_for`. `wait_for` can now take any number of futures or a condition block. This does need to be a block returning a boolean value since it must be evaluated each time through.

antstorm commented on lib/temporal/workflow/context.rb:
```diff
@@ -215,14 +215,53 @@ def wait_for_all(*futures)
+          else
+            should_yield = true
+
+            dispatcher.register_handler(Dispatcher::WILDCARD, Dispatcher::WILDCARD) do
```
There's a slight problem here — we're not de-registering the handler afterwards, which means that it will be called for every event until the end of a workflow. It won't be doing anything because of that `blocked` variable, but something we might want to address. It wasn't an issue previously because each target was expected to fire each event only once, but still keeps these references in memory.

This can be done in a follow-up, I might have a look at it

jeffschoner-stripe replied:
I suppose once it's unblocked, it can be unregistered. The interplay on this is messy because it needs a callback. I'll see if I can do this cleanly or if it's best to revise later.

antstorm commented on lib/temporal/workflow/context.rb:
```diff
@@ -215,14 +215,53 @@ def wait_for_all(*futures)
+          else
+            should_yield = true
+            futures.each do |future|
+              dispatcher.register_handler(future.target, Dispatcher::WILDCARD) do
```
I wonder if we actually need to separate between the futures and unblock condition cases here? There are probably arguments to both approaches:

1. Only using a single wildcard handler for both cases means that it might get called more than needed, but it simplifies the code (only need to register one handler)
2. Using future-specific and a wildcard handlers is a bit of an optimisation for when only futures are passed, but is actually less optimal when both are passed

I wonder what do you think about this one?

jeffschoner-stripe replied:
I believe this code path is used when invoking activities synchronously or sleeps, which are almost certainly the most common cases. I could register separate handlers here, rather than handling the separate cases inside the dispatcher. I'll give this alternative a try to see how it looks once implemented, especially with the unregistration in place.

antstorm replied:
Sounds good 👍 But also happy to rearrange things in later PRs to keep this one contained and ship the feature

antstorm commented on lib/temporal/workflow/context.rb:
```diff
@@ -215,14 +215,53 @@ def wait_for_all(*futures)
+        blocked = true
+
+        if futures.any?
+          if futures.all?(&:finished?)
```
This should probably be `futures.any?(&:finished?)`, right? At least judging by the example and a test implementation it waits for a single future before proceeding

jeffschoner-stripe replied:
Found another similar issue here where if the condition already evaluates to true, we should not block on the futures, yield, or register any dispatch handlers for them

antstorm reviewed (commented):
Looks great, just a few more wrinkles to iron out, but it's almost there 👍

jeffschoner-stripe commented:
I've fixed the `any?`/`all?` mix-up, as well as a related case, and covered both in the integration test. Agree saving unregistration for a subsequent PR is worthwhile to improve performance.

jeffschoner-stripe commented:
I've also tidied my commits for a cleaner merge

antstorm reviewed (approved):
Looks great, thank you @jeffschoner-stripe, this is a big step forward!
